File: agent/src/main/java/com/octogonapus/omj/agent/OMJClassAdapter.java
/*
 * This file is part of OMJ.
 *
 * OMJ is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OMJ is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OMJ.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.octogonapus.omj.agent;

import java.util.Arrays;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public final class OMJClassAdapter extends ClassVisitor implements Opcodes {

  private final Logger logger = LoggerFactory.getLogger(OMJClassAdapter.class);
  private final DynamicClassDefiner dynamicClassDefiner;
  private final ClassFilter classFilter;
  private int classVersion;
  private String className;
  private String superName;

  public OMJClassAdapter(
      final int api,
      final ClassVisitor classVisitor,
      final DynamicClassDefiner dynamicClassDefiner,
      final ClassFilter classFilter) {
    super(api, classVisitor);
    this.dynamicClassDefiner = dynamicClassDefiner;
    this.classFilter = classFilter;
  }

  @Override
  public void visit(
      final int version,
      final int access,
      final String name,
      final String signature,
      final String superName,
      final String[] interfaces) {
    logger.debug(
        "version = {}, access = {}, name = {}, signature = {}, superName = {}, interfaces = {}",
        version,
        access,
        name,
        signature,
        superName,
        Arrays.deepToString(interfaces));

    super.visit(version, access, name, signature, superName, interfaces);
    classVersion = version;
    className = name;
    this.superName = superName;
  }

  @Override
  public MethodVisitor visitMethod(
      final int access,
      final String name,
      final String descriptor,
      final String signature,
      final String[] exceptions) {
    logger.debug(
        "access = {}, name = {}, descriptor = {}, signature = {}, exceptions = {}",
        access,
        name,
        descriptor,
        signature,
        Arrays.deepToString(exceptions));

    final var visitor = super.visitMethod(access, name, descriptor, signature, exceptions);

    if (isInstanceInitializationMethod(name, descriptor)) {
      return new OMJInstanceInitializationMethodAdapter(
          api, visitor, dynamicClassDefiner, classFilter, descriptor, className, superName);
    } else if (isClassInitializationMethod(name, descriptor, access)) {
      return new OMJMethodAdapter(
          api,
          visitor,
          dynamicClassDefiner,
          classFilter,
          descriptor,
          hasAccessFlag(access, ACC_STATIC),
          className);
    } else if (isMainMethod(name, descriptor, access)) {
      return new OMJMainMethodAdapter(api, visitor, dynamicClassDefiner, classFilter, className);
    } else {
      return new OMJMethodAdapter(
          api,
          visitor,
          dynamicClassDefiner,
          classFilter,
          descriptor,
          hasAccessFlag(access, ACC_STATIC),
          className);
    }
  }

  /**
   * Determines whether the method is the "main method" (entry point) according to JLS Section
   * 12.1.4.
   *
   * @param methodName The name of the method.
   * @param descriptor The method's descriptor.
   * @param access The method's access number.
   * @return True if the method is the "main method".
   */
  private boolean isMainMethod(final String methodName, final String descriptor, final int access) {
    final Type[] argumentTypes = Type.getArgumentTypes(descriptor);
    return methodName.equals("main")
        && hasAccessFlag(access, ACC_PUBLIC)
        && hasAccessFlag(access, ACC_STATIC)
        && Type.getReturnType(descriptor).getSort() == Type.VOID
        && argumentTypes.length == 1
        && argumentTypes[0].getDescriptor().equals("[Ljava/lang/String;");
  }

  /**
   * Determines whether the method is an instance initialization method according to JVMS Section
   * 2.9.1.
   *
   * @param methodName The method's name.
   * @param descriptor The method's descriptor.
   * @return True if the method is an instance initialization method.
   */
  private boolean isInstanceInitializationMethod(final String methodName, final String descriptor) {
    return methodName.equals("<init>") && Type.getReturnType(descriptor).getSort() == Type.VOID;
  }

  /**
   * Determines whether the method is an class initialization method according to JVMS Section
   * 2.9.2.
   *
   * @param methodName The method's name.
   * @param descriptor The method's descriptor.
   * @return True if the method is an class initialization method.
   */
  private boolean isClassInitializationMethod(
      final String methodName, final String descriptor, final int access) {
    final int majorVersion = classVersion & 0xFFFF;

    final boolean versionCheck;
    if (majorVersion >= 51) {
      versionCheck =
          hasAccessFlag(access, ACC_STATIC) && Type.getArgumentTypes(descriptor).length == 0;
    } else {
      versionCheck = true;
    }

    return methodName.equals("<clinit>")
        && Type.getReturnType(descriptor).getSort() == Type.VOID
        && versionCheck;
  }

  /**
   * Checks if an access flag is present. See {@link Opcodes} for the flags.
   *
   * @param access The access int to check.
   * @param flag The flag to check for.
   * @return True if the flag is present.
   */
  private static boolean hasAccessFlag(final int access, final int flag) {
    return (access & flag) == flag;
  }
}

File: agent/src/main/kotlin/com/octogonapus/omj/agent/OMJClassAdapter.kt
/*
 * This file is part of OMJ.
 *
 * OMJ is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * OMJ is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with OMJ.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.octogonapus.omj.agent

import com.octogonapus.omj.di.OMJKoinComponent
import com.octogonapus.omj.util.nullableSingleAssign
import com.octogonapus.omj.util.singleAssign
import mu.KotlinLogging
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type

internal class OMJClassAdapter(
    api: Int,
    classVisitor: ClassVisitor?
) : ClassVisitor(api, classVisitor), OMJKoinComponent {

    private var classVersion = 0
    private var className by singleAssign<String>()
    private var superName by nullableSingleAssign<String>()

    override fun visit(
        version: Int,
        access: Int,
        name: String,
        signature: String?,
        superName: String?,
        interfaces: Array<String>?
    ) {
        logger.debug {
            """
            version = $version
            access = $access
            name = $name
            signature = $signature
            superName = $superName
            interfaces = ${interfaces?.contentDeepToString()}
            """.trimIndent()
        }

        classVersion = version
        className = name
        this.superName = superName

        super.visit(version, access, name, signature, superName, interfaces)
    }

    override fun visitMethod(
        access: Int,
        name: String,
        descriptor: String,
        signature: String?,
        exceptions: Array<String>?
    ): MethodVisitor {
        logger.debug {
            """
            access = $access
            name = $name
            descriptor = $descriptor
            signature = $signature
            exceptions = ${exceptions?.contentDeepToString()}
            """.trimIndent()
        }

        val visitor = super.visitMethod(access, name, descriptor, signature, exceptions)

        // If superName is null, we are visiting the Object class, so there is nothing for us to
        // instrument. Otherwise, determine which adapter to use.
        return superName?.let { superName ->
            determineMethodAdapter(name, descriptor, visitor, superName, access)
        } ?: visitor
    }

    private fun determineMethodAdapter(
        name: String,
        descriptor: String,
        visitor: MethodVisitor,
        superName: String,
        access: Int
    ): MethodVisitor {
        val method = Method(name, descriptor)

        return when {
            isInstanceInitializationMethod(method) ->
                OMJInstanceInitializationMethodAdapter(
                    api,
                    visitor,
                    method,
                    className,
                    superName
                )

            isClassInitializationMethod(method, access) -> OMJMethodAdapter(
                api,
                visitor,
                method,
                hasAccessFlag(access, Opcodes.ACC_STATIC),
                className
            )

            isMainMethod(method, access) -> OMJMainMethodAdapter(api, visitor, method, className)

            else -> OMJMethodAdapter(
                api,
                visitor,
                method,
                hasAccessFlag(access, Opcodes.ACC_STATIC),
                className
            )
        }
    }

    /**
     * Determines whether the method is the "main method" (entry point) according to JLS Section
     * 12.1.4.
     *
     * @param method The method to check.
     * @param access The method's access number.
     * @return True if the method is the "main method".
     */
    private fun isMainMethod(method: Method, access: Int): Boolean {
        val argumentTypes = Type.getArgumentTypes(method.descriptor)
        return (
            method.name == "main" && hasAccessFlag(access, Opcodes.ACC_PUBLIC) &&
                hasAccessFlag(access, Opcodes.ACC_STATIC) &&
                Type.getReturnType(method.descriptor).sort == Type.VOID &&
                argumentTypes.size == 1 &&
                argumentTypes[0].descriptor == "[Ljava/lang/String;"
            )
    }

    /**
     * Determines whether the method is an instance initialization method according to JVMS Section
     * 2.9.1.
     *
     * @param method The method to check.
     * @return True if the method is an instance initialization method.
     */
    private fun isInstanceInitializationMethod(method: Method) =
        method.name == "<init>" && Type.getReturnType(method.descriptor).sort == Type.VOID

    /**
     * Determines whether the method is an class initialization method according to JVMS Section
     * 2.9.2.
     *
     * @param method The method to check.
     * @return True if the method is an class initialization method.
     */
    private fun isClassInitializationMethod(method: Method, access: Int): Boolean {
        val majorVersion = classVersion and 0xFFFF

        val versionCheck = if (majorVersion >= 51) {
            hasAccessFlag(access, Opcodes.ACC_STATIC) &&
                Type.getArgumentTypes(method.descriptor).isEmpty()
        } else {
            true
        }

        return method.name == "<clinit>" &&
            Type.getReturnType(method.descriptor).sort == Type.VOID &&
            versionCheck
    }

    companion object {

        private val logger = KotlinLogging.logger { }

        /**
         * Checks if an access flag is present. See [Opcodes] for the flags.
         *
         * @param access The access int to check.
         * @param flag The flag to check for.
         * @return True if the flag is present.
         */
        private fun hasAccessFlag(access: Int, flag: Int) = access and flag == flag
    }
}
